src/lib/process/process_win32.c

```diff
@@ -474,28 +474,43 @@ process_win32_timer_callback(periodic_timer_t *timer, void *data)
   tor_assert(timer == periodic_timer);
   tor_assert(data == NULL);
 
-  log_debug(LD_PROCESS, "Windows Process I/O timer ticked");
-
   /* Move the process into an alertable state. */
   process_win32_trigger_completion_callbacks();
 
   /* Check if our processes are still alive. */
-  const smartlist_t *processes = process_get_all_processes();
 
-  SMARTLIST_FOREACH(processes, process_t *, p,
-                    process_win32_timer_test_process(p));
+  /* Since the call to process_win32_timer_test_process() might call
+   * process_notify_event_exit() which again might call process_free() which
+   * updates the list of processes returned by process_get_all_processes() it
+   * is important here that we make sure to not touch the list of processes if
+   * the call to process_win32_timer_test_process() returns true. */
+  bool done = true;
+
+  do {
+    const smartlist_t *processes = process_get_all_processes();
+
+    SMARTLIST_FOREACH_BEGIN(processes, process_t *, process) {
+      if (process_win32_timer_test_process(process)) {
+        done = false;
+        break;
```

Comment:
I looked and saw that if `true` is returned, then it means the exit callback was called thus the `process` was freed at that point.

1) I would comment out here that it is imperative that if `true` is returned, the loop must end (we can't continue in the list because a `smartlist_remove()` occurred.

2) Why is `done = false` when we break out of the smartlist loop and then the first thing that is done is `done = true` ?